### ISSUE TITLE
VSCode terminal seems to construct environment in a strange way

### ISSUE THREAD

**Issue by bakhil** [microsoft/vscode#205013]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.86.1 (commit 31c37ee8f63491495ac49e43b8544550fbae4533)
- OS Version: WSL2 (issue seems to reproduce on linux as well)

I initially posted this on [stackoverflow](https://stackoverflow.com/q/77978975/8278183) but after some more observations, it looks like a bug to me. First, here is the basic issue. I am starting vscode using the cli web server as `code serve-web`. I am using the following in `settings.json`:
```
{
    "terminal.integrated.profiles.linux": {
        "my-bash": {
            "path": "/usr/bin/bash",
            "args": ["--noprofile", "--norc"]
        },
    },
    "terminal.integrated.defaultProfile.linux": "my-bash"
}
```

To replicate my issue, put `export TMP_VARIABLE=10` in `/etc/profile.d/xxxx.sh` and run the following commands (on WSL/linux, not the vscode terminal).
```
$ export TMP_VARIABLE=3
$ export NEW_TMP_VARIABLE=100
$ code serve-web
```
In the vscode terminal, the newly created variable `NEW_TMP_VARIABLE` appears and is set to `100` (which is what I want). But for some reason, `TMP_VARIABLE` is still equal to `10`. Why is this happening?

How do I get vscode web server's terminal to use the variables in my current environment? And why does the `--noprofile` option not have the desired effect?

---

**Further observations**

If I add `echo 'profile.d is being read!!'` to `/etc/profile.d/xxxx.sh`, this statement is output each time I open a new terminal on WSL/linux. In vscode, if I change the shell args to `["--login"]`, I can see it output in the vscode terminal as well. But with args set to `["--noprofile", "--norc"]`, I don't see this echo. However, the value of `TMP_VARIABLE` is still set according to what's in `/etc/profile.d/xxxx.sh`, not according to the local environment. This makes me think vscode is starting a new login shell somewhere and copying its environment over in some strange way. I'm not sure, but seems to me like there is some bug here.

**Comment by meganrogge** [comment]
does this reproduce if you disable shell integration? `terminal.integrated.shellIntegration.enabled`?

**Comment by bakhil** [comment]
Yes, it looks like the issue is still there after `terminal.integrated.shellIntegration.enabled` is set to `false`. My `settings.json` is the following now:
```
{
    "terminal.integrated.profiles.linux": {
        "my-bash": {
            "path": "/usr/bin/bash",
            "args": ["--noprofile", "--norc"]
        },
    },
    "terminal.integrated.defaultProfile.linux": "my-bash",
    "terminal.integrated.shellIntegration.enabled": false
}
```
The shell has `TMP_VARIABLE` from `/etc/profile.d/xxxx.sh` even though I `export`ed a different value before calling `code serve-web`.

**Comment by connor4312** [comment]
serve-web doesn't do anything magic, this looks like VS Code's default shell environment resolution

**Comment by bpasero** [comment]
I am not in the position to fully understand how it works in this kind of setup, but when you start VSCode desktop normally from a terminal, all environment variables are inherited to all child processes.

Here it seems a combination of remote, WSL/Linux and web server are involved where I am not sure if we do some kind of environment variables cleansing on the way.

**Comment by aeschli** [comment]
I suspect its coming from https://github.com/microsoft/vscode/blob/df4af4aefd2348eb20dbd2bdc9322042fa71733c/src/vs/platform/shell/node/shellEnv.ts#L35

You can try to start the server with `--force-disable-user-env`, but it seems the cli does not forward that argument

**Comment by bakhil** [comment]
So.. is there any plan to fix this in the near future? Otherwise we can probably set up some redirecting script that stores the current environment to a local file whenever `code serve-web` is called and reads from that when a terminal is opened.

**Comment by bpasero** [comment]
Does it help if you configure the `terminal.integrated.inheritEnv` setting to not inherit the environment? I think we still resolve the shell environment for each new terminal here:

https://github.com/microsoft/vscode/blob/01ce8d869e4220577e194b2dd57bfa12437dfcf5/src/vs/server/node/remoteTerminalChannel.ts#L203

@Tyriar @meganrogge would this override variables configured by the user?

**Comment by Tyriar** [comment]
I should have provided more thoughts when reassigning, but I was guessing this was related to server CLI lifecycle; @connor4312 could an older server process have been around like how the main process sticks around?

**Comment by connor4312** [comment]
Yes, multiple server processes could exist concurrently. Whenever a user loads the URL of code-server we try to look up the latest server version, then serve that to the client. Multiple clients(/browser tabs, in the common case) can be connected at once with multiple server versions running if done over time.
